python/ray/serve/_private/proxy_router.py

```diff
@@ -126,6 +143,29 @@ def match_route(
 
         return None
 
+    def ready_for_traffic(self, is_head: bool) -> Tuple[bool, str]:
+        """Whether the proxy router is ready to serve traffic.
+
+        The first return value will be false if any of the following hold:
+        - The route table has not been populated yet
+        - The route table has been populated, but none of the handles
+          have received running replicas yet AND it lives on a worker node.
+
+        Otherwise, the first return value will be true.
+        """
+
+        if not self._route_table_populated:
+            return False, NO_ROUTES_MESSAGE
+
+        if is_head:
+            return True, ""
+
+        for handle in self.handles.values():
+            if handle.running_replicas_populated():
```

Comment:
this name is misleading -- it could be that the running replicas have been populated at least once, but the size of the list is `0`

suggest to call it something like `has_running_replicas`

Reply:
hmm so right now what I've implemented is that the proxy is considered ready if it's received a non-zero list of replicas at least once, but if after that the number of replicas scale back down to 0, the flag isn't flipped off. For this I think `running_replicas_populated` is probably still more accurate. Implementing it this way targets more of the "proxy just came up and hasn't had a chance to receive replicas yet before head node goes down" case.

I can also change it so that if the number of replicas scale back down to 0, the flag is flipped off -> the proxy changes to unhealthy so that it won't receive traffic. This will cover more bases, but will put more pressure on the head node proxy when all ingress deployments have scaled to zero. WDYT?

Reply:
Got it. I think the existing is reasonable then.